### PR TITLE
Clean up cullOverReferences pass

### LINKED ISSUE #14902: Tuple expressions appear to ignore constness when capturing elements
### Summary of Problem

The CHIP6 for tuples claims that if the default argument intent of a type is `ref` or `const ref`, then that element will be captured into the tuple by `ref`. However that seems to lead to strange/ arguably incorrect behavior when capturing a `const ref` alias into a tuple expression.

It is possible to capture a `const ref` alias into a tuple by `ref`, which allows you to modify the original value.

Ostensibly I would expect this to be an error.

### Steps to Reproduce

**Source Code:**

```chapel
record someRecord { var x: int = 0; }

var rec: someRecord;

const ref foo = rec;

// Shouldn't this be illegal, since "foo" is a `const ref`?
(foo,)[1].x = 128;

writeln(foo.x:string);
```

Outputs `128`, when I would hope for some sort of compiler error.

**Compile command:**

`chpl foo.chpl`

**Execution command:**

`./foo`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR implements a "light" refactor of the `cullOverReferences` compiler pass. As part of my work to resolve #14902 I've had to study the code in `cullOverReferences`. I found it hard to understand the structure of the pass because it was composed of a monolithic "block" of code.

- Add a context object used to store the state of the pass.
- Move each step of the pass into a method on the context.
- Split the body of the 600~ line for-loop iterating over collected symbols into separate functions, each prefaced with `check` for consistency.
- Reduce nesting in `check` methods by inverting non-cast conditionals on AST to exit early (i.e. `if (call->isPrimitive(PRIM_MOVE))` becomes `if (!call->isPrimitive(...) return false;`).
- Moves and reformats existing comments as necessary.

Also add a few futures capturing issues with tuples.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

File: compiler/resolution/cullOverReferences.cpp [hunk 1]
```diff
@@ -582,746 +582,848 @@ bool isRefOrTupleWithRef(Symbol* index, int tupleElement)
...
+  bool checkCompilerRefTemporaries(CallExpr*, GraphNode, bool&);
+  bool checkOuterVariableInForall(SymExpr*, GraphNode, bool&);
+  void visitCollectedSymbol(GraphNode);
+
+  void propagateNonConstnessThroughRevisitGraph(void);
+  void markRemainingSymsWithUnknownConstnessConst(void);
+  void lowerRemainingContextCallExprs(void);
+  void performLateConstCheck(void);
+
+private:
```
Comment (mppf) [inline compiler/resolution/cullOverReferences.cpp:624]: I don't personally see the point of making anything private for a class that only can be used within one file, but that's just a style thing

File: compiler/resolution/cullOverReferences.cpp [hunk 2]
```diff
@@ -582,746 +582,848 @@ bool isRefOrTupleWithRef(Symbol* index, int tupleElement)
...
-     "header file" in addition to "implementation files" as part of a library,
-     there is no problem.
-
-     (However, any interface describing a callable function / call-back
-      function will need to have some other default -- possibly blank
-      argument intent leads to compilation error for ref-if-modified types
-      like arrays).
+// TODO: Move this into `resolution/loopDetails.cpp` and adjust the function
+// `gatherLoopDetails` to use it.
+struct LoopInfo {
```
Comment (mppf) [inline compiler/resolution/cullOverReferences.cpp:590]: It is sometimes necessary to use anonymous namespaces when declaring types meant to be used only within one .cpp file. It's not a thing you must do here but I wanted to make you aware of the potential for an issue. If another .cpp defines a (local) struct with the same name, there will be link errors, unless each uses `namespace { }` around the type definitions.

File: compiler/resolution/cullOverReferences.cpp (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/resolution/cullOverReferences.cpp:624]: Yeah, I couldn't make up my mind on whether to make it a class or a struct. I'll keep that in mind for next time.

File: compiler/resolution/cullOverReferences.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/resolution/cullOverReferences.cpp:590]: Ok. In the future I would like to migrate this to be used in `gatherLoopDetails` at which point I could put it in a header file.

Comment (mppf) [reply compiler/resolution/cullOverReferences.cpp:590]: Right, but the same thing applies to the new type just below.

**Comment by dlongnecke-cray** [comment]
Waiting to merge this due to problems with `test/performance/ferguson/array-assign-block.chpl` seeing some segfaults on gasnet with LLVM.

**Comment by dlongnecke-cray** [comment]
Segfaults in `performance/ferguson` are very likely unrelated, going to merge.

**Pull request merged**